ast/walk.go

```diff
@@ -0,0 +1,74 @@
+package ast
+
+import (
+	"iter"
+)
+
+//go:generate go run ../tools/gen-ast-walk/main.go -astfile ast.go -constfile ast_const.go -outfile walk_internal.go
+
+// Visitor is an interface for visiting AST nodes.
+// If the result of Visit is nil, the node will not be traversed.
+type Visitor interface {
+	Visit(node Node) Visitor
+	Field(name string) Visitor
+	Index(index int) Visitor
+}
+
+type stackItem struct {
+	node    Node
+	visitor Visitor
+}
+
+// Walk traverses an AST in depth-first order.
+func Walk(node Node, v Visitor) {
+	var stack []*stackItem
+	stack = append(stack, &stackItem{node: node, visitor: v})
+
+	for len(stack) > 0 {
+		last := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+		if last.node == nil {
+			continue
+		}
+
+		v := last.visitor.Visit(last.node)
+		if v == nil {
+			continue
+		}
+
+		stack = walkInternal(last.node, v, stack)
+	}
+}
+
+type inspector func(Node) bool
+
+func (f inspector) Visit(node Node) Visitor {
+	if f(node) {
+		return f
+	}
+	return nil
+}
+
+func (f inspector) Field(name string) Visitor {
+	return f
+}
+
+func (f inspector) Index(index int) Visitor {
+	return f
+}
+
+// Inspect traverses an AST in depth-first order and calls f for each node.
+func Inspect(node Node, f func(Node) bool) {
+	Walk(node, inspector(f))
+}
+
+// Preorder returns an iterator that traverses an AST in depth-first preorder.
+func Preorder(node Node) iter.Seq[Node] {
```

Comment:
Note: `go.mod` still have `go 1.22.0` dilective so we can't use range over func in memefish development.

Reply:
`go` directive in `go.mod` is also updated. Is it problematic?

Reply:
Yeah, seems like I had missed this.